File: examples/python/batch_computation.py
"""
Batch Computation Examples

This example demonstrates how to use batch computation functions (pdist and cdist)
to efficiently calculate distances between multiple trajectories.
"""

import numpy as np
import traj_dist_rs

print("=" * 60)
print("Batch Computation Examples")
print("=" * 60)

# Create a set of sample trajectories
num_trajectories = 10
trajectories = [
    np.random.rand(20, 2) * 10  # Each trajectory has 20 points in [0, 10] range
    for _ in range(num_trajectories)
]

print(f"\nCreated {num_trajectories} trajectories with 20 points each")

# Method 1: Using pdist (pairwise distances in compressed format)
print("\n1. pdist - Pairwise Distance Matrix (Compressed)")
print("-" * 60)

# Create a Metric object for pdist
metric = traj_dist_rs.Metric.sspd(type_d="euclidean")

# Calculate pairwise distances using pdist
# Returns compressed distance matrix (upper triangle in row-major order)
compressed_distances = traj_dist_rs.pdist(trajectories, metric)

print(f"  Number of pairwise distances: {len(compressed_distances)}")
print(f"  Expected: {num_trajectories * (num_trajectories - 1) // 2}")
print(f"  First 5 distances: {compressed_distances[:5]}")
print(f"  Min distance: {compressed_distances.min():.6f}")
print(f"  Max distance: {compressed_distances.max():.6f}")

# Method 2: Using pdist with different algorithms
print("\n2. pdist with Different Algorithms")
print("-" * 60)

# Create metrics using factory methods
metrics = [
    ("SSPD", traj_dist_rs.Metric.sspd(type_d="euclidean")),
    ("DTW", traj_dist_rs.Metric.dtw(type_d="euclidean")),
    ("Hausdorff", traj_dist_rs.Metric.hausdorff(type_d="euclidean")),
    ("LCSS", traj_dist_rs.Metric.lcss(eps=0.1, type_d="euclidean")),
    ("EDR", traj_dist_rs.Metric.edr(eps=0.1, type_d="euclidean")),
    ("ERP", traj_dist_rs.Metric.erp(g=[0.0, 0.0], type_d="euclidean")),
    ("Discret Frechet", traj_dist_rs.Metric.discret_frechet(type_d="euclidean")),
]

for name, metric in metrics:
    try:
        distances = traj_dist_rs.pdist(trajectories, metric)
        print(
            f"  {name:20s}: {len(distances)} distances, " f"mean={distances.mean():.6f}"
        )
    except Exception as e:
        print(f"  {name:20s}: Error - {e}")

# Method 3: Using cdist (full distance matrix)
print("\n3. cdist - Full Distance Matrix")
print("-" * 60)

# Create two sets of trajectories
traj_set_a = trajectories[:5]  # First 5 trajectories
traj_set_b = trajectories[5:]  # Last 5 trajectories

# Calculate distance matrix using cdist
metric = traj_dist_rs.Metric.sspd(type_d="euclidean")
distance_matrix = traj_dist_rs.cdist(traj_set_a, traj_set_b, metric)

print(f"  Distance matrix shape: {distance_matrix.shape}")
print(f"  Expected: ({len(traj_set_a)}, {len(traj_set_b)})")
print("  Full matrix:")
print(distance_matrix)

# Method 4: cdist with spherical distance
print("\n4. cdist with Spherical Distance")
print("-" * 60)

# Create trajectories with geographic coordinates (latitude, longitude)
geo_trajectories = [
    np.array(
        [
            [40.7128, -74.0060],  # New York
            [40.7306, -73.9352],  # LaGuardia
            [40.6413, -73.7781],  # JFK
        ]
    ),
    np.array(
        [
            [34.0522, -118.2437],  # Los Angeles
            [34.0522, -118.2437],  # LAX (approximate)
            [33.9425, -118.4081],  # Another LA area point
        ]
    ),
    np.array(
        [
            [51.5074, -0.1278],  # London
            [51.4700, -0.4543],  # Heathrow
            [51.5055, -0.2799],  # Another London area point
        ]
    ),
]

metric = traj_dist_rs.Metric.sspd(type_d="spherical")
distance_matrix = traj_dist_rs.cdist(geo_trajectories, geo_trajectories, metric)

print("  Distance matrix (Haversine, in kilometers):")
print(distance_matrix / 1000)  # Convert to kilometers

# Method 5: Using Metric with parameters
print("\n5. pdist with Metric Parameters")
print("-" * 60)

# LCSS with epsilon parameter
metric_lcss = traj_dist_rs.Metric.lcss(eps=0.1, type_d="euclidean")
distances_lcss = traj_dist_rs.pdist(trajectories, metric_lcss)
print(f"  LCSS (epsilon=0.1): {distances_lcss.mean():.6f}")

# EDR with epsilon parameter
metric_edr = traj_dist_rs.Metric.edr(eps=0.2, type_d="euclidean")
distances_edr = traj_dist_rs.pdist(trajectories, metric_edr)
print(f"  EDR (epsilon=0.2): {distances_edr.mean():.6f}")

print("\n" + "=" * 60)
print("All batch computations completed successfully!")
print("=" * 60)

File: examples/batch_computation.rs
//! Batch Computation Examples
//!
//! This example demonstrates how to use batch computation functions (pdist and cdist)
//! to efficiently calculate distances between multiple trajectories.

use traj_dist_rs::distance::{
    batch::{DistanceAlgorithm, Metric, cdist, pdist},
    distance_type::DistanceType,
};

fn main() {
    println!("============================================================");
    println!("Batch Computation Examples");
    println!("============================================================");

    // Create a set of sample trajectories
    let num_trajectories = 10;
    let trajectories: Vec<Vec<[f64; 2]>> = (0..num_trajectories)
        .map(|_| {
            (0..20)
                .map(|_| [rand::random::<f64>() * 10.0, rand::random::<f64>() * 10.0])
                .collect()
        })
        .collect();

    println!(
        "\nCreated {} trajectories with 20 points each",
        num_trajectories
    );

    // Method 1: Using pdist (pairwise distances in compressed format)
    println!("\n1. pdist - Pairwise Distance Matrix (Compressed)");
    println!("{}", "-".repeat(60));

    // Calculate pairwise distances using pdist
    let metric = Metric::new(DistanceAlgorithm::SSPD, DistanceType::Euclidean);
    match pdist(&trajectories, &metric, false) {
        Ok(compressed_distances) => {
            println!(
                "  Number of pairwise distances: {}",
                compressed_distances.len()
            );
            println!(
                "  Expected: {}",
                num_trajectories * (num_trajectories - 1) / 2
            );
            println!("  First 5 distances:");
            for (i, &dist) in compressed_distances.iter().take(5).enumerate() {
                println!("    [{}]: {:.6}", i, dist);
            }

            let min_dist = compressed_distances
                .iter()
                .cloned()
                .fold(f64::INFINITY, f64::min);
            let max_dist = compressed_distances
                .iter()
                .cloned()
                .fold(f64::NEG_INFINITY, f64::max);
            println!("  Min distance: {:.6}", min_dist);
            println!("  Max distance: {:.6}", max_dist);
        }
        Err(e) => {
            println!("  Error: {}", e);
        }
    }

    // Method 2: Using pdist with different algorithms
    println!("\n2. pdist with Different Algorithms");
    println!("{}", "-".repeat(60));

    let algorithms = vec![
        (DistanceAlgorithm::SSPD, "SSPD"),
        (DistanceAlgorithm::DTW, "DTW"),
        (DistanceAlgorithm::Hausdorff, "Hausdorff"),
        (DistanceAlgorithm::LCSS { eps: 0.1 }, "LCSS"),
        (DistanceAlgorithm::EDR { eps: 0.1 }, "EDR"),
        (DistanceAlgorithm::ERP { g: [0.0, 0.0] }, "ERP"),
        (DistanceAlgorithm::DiscretFrechet, "Discret Frechet"),
    ];

    for (algorithm, name) in algorithms {
        let metric = Metric::new(algorithm, DistanceType::Euclidean);
        match pdist(&trajectories, &metric, false) {
            Ok(distances) => {
                let mean = distances.iter().sum::<f64>() / distances.len() as f64;
                println!(
                    "  {:20}: {} distances, mean={:.6}",
                    name,
                    distances.len(),
                    mean
                );
            }
            Err(e) => {
                println!("  {:20}: Error - {}", name, e);
            }
        }
    }

    // Method 3: Using cdist (full distance matrix)
    println!("\n3. cdist - Full Distance Matrix");
    println!("{}", "-".repeat(60));

    // Create two sets of trajectories
    let traj_set_a: Vec<_> = trajectories.iter().take(5).cloned().collect();
    let traj_set_b: Vec<_> = trajectories.iter().skip(5).cloned().collect();

    // Calculate distance matrix using cdist
    let metric = Metric::new(DistanceAlgorithm::SSPD, DistanceType::Euclidean);
    match cdist(&traj_set_a, &traj_set_b, &metric, false) {
        Ok(distance_matrix) => {
            let rows = traj_set_a.len();
            let cols = traj_set_b.len();
            println!("  Distance matrix shape: {}x{}", rows, cols);
            println!("  Expected: {}x{}", rows, cols);
            println!("  Full matrix:");
            for row in 0..rows {
                print!("    [");
                for col in 0..cols {
                    let idx = row * cols + col;
                    if col > 0 {
                        print!(", ");
                    }
                    print!("{:.6}", distance_matrix[idx]);
                }
                println!("]");
            }
        }
        Err(e) => {
            println!("  Error: {}", e);
        }
    }

    // Method 4: cdist with spherical distance
    println!("\n4. cdist with Spherical Distance");
    println!("{}", "-".repeat(60));

    // Create trajectories with geographic coordinates (latitude, longitude)
    let geo_trajectories = vec![
        vec![
            [40.7128, -74.0060],
            [40.7306, -73.9352],
            [40.6413, -73.7781],
        ], // New York
        vec![
            [34.0522, -118.2437],
            [34.0522, -118.2437],
            [33.9425, -118.4081],
        ], // Los Angeles
        vec![[51.5074, -0.1278], [51.4700, -0.4543], [51.5055, -0.2799]], // London
    ];

    let metric = Metric::new(DistanceAlgorithm::SSPD, DistanceType::Spherical);
    match cdist(&geo_trajectories, &geo_trajectories, &metric, false) {
        Ok(distance_matrix) => {
            let rows = geo_trajectories.len();
            let cols = geo_trajectories.len();
            println!("  Distance matrix (Haversine, in kilometers):");
            for row in 0..rows {
                print!("    [");
                for col in 0..cols {
                    let idx = row * cols + col;
                    if col > 0 {
                        print!(", ");
                    }
                    print!("{:.3}", distance_matrix[idx] / 1000.0); // Convert to kilometers
                }
                println!("]");
            }
        }
        Err(e) => {
            println!("  Error: {}", e);
        }
    }

    // Method 5: pdist with parallel processing
    println!("\n5. pdist with Parallel Processing");
    println!("{}", "-".repeat(60));

    let metric = Metric::new(DistanceAlgorithm::SSPD, DistanceType::Euclidean);

    // Sequential
    match pdist(&trajectories, &metric, false) {
        Ok(distances) => {
            println!("  Sequential: {} distances computed", distances.len());
        }
        Err(e) => {
            println!("  Sequential error: {}", e);
        }
    }

    // Parallel
    match pdist(&trajectories, &metric, true) {
        Ok(distances) => {
            println!("  Parallel: {} distances computed", distances.len());
        }
        Err(e) => {
            println!("  Parallel error: {}", e);
        }
    }

    // Method 6: cdist with parallel processing
    println!("\n6. cdist with Parallel Processing");
    println!("{}", "-".repeat(60));

    let metric = Metric::new(DistanceAlgorithm::DTW, DistanceType::Euclidean);

    // Sequential
    match cdist(&traj_set_a, &traj_set_b, &metric, false) {
        Ok(_matrix) => {
            let rows = traj_set_a.len();
            let cols = traj_set_b.len();
            println!("  Sequential: {}x{} matrix computed", rows, cols);
        }
        Err(e) => {
            println!("  Sequential error: {}", e);
        }
    }

    // Parallel
    match cdist(&traj_set_a, &traj_set_b, &metric, true) {
        Ok(_matrix) => {
            let rows = traj_set_a.len();
            let cols = traj_set_b.len();
            println!("  Parallel: {}x{} matrix computed", rows, cols);
        }
        Err(e) => {
            println!("  Parallel error: {}", e);
        }
    }

    println!("\n============================================================");
    println!("All batch computations completed successfully!");
    println!("============================================================");
}

// Note: This example requires the `rand` crate for random number generation.
// Add this to your Cargo.toml:
// [dependencies]
// rand = "0.8"
// traj-dist-rs = { path = "..", features = ["parallel"] }
